AX: [Site Isolation] iOS remote element bring-up

hoffmanjoshua opened the pull request:
#### 93d7ed9bebe7dd7bbebdbe6f6c062e0994a5e91d
<pre>
AX: [Site Isolation] iOS remote element bring-up
<a href="https://bugs.webkit.org/show_bug.cgi?id=284847">https://bugs.webkit.org/show_bug.cgi?id=284847</a>
<a href="https://rdar.apple.com/problem/141640442">rdar://problem/141640442</a>

Reviewed by Tyler Wilcock.

This patch brings-up site isolation support for iOS accessibility. This involves
setting up AXRemoteFrames properly, using existing stub methods/IPC messages
that are used on macOS. After this, WebKit can create a remote element child
that binds the two processes with a matching token.

WebKit::newAccessibilityRemoteToken was moved to WebCore, so that it may be
used in AccessibilityObjectIOS.

Note: this patch does not address hit testing. That will be resolved in a follow
up.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/ios/AXRuntimeSPI.h: Added.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::addRemoteFrameChild):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::unarchivedTokenForData):
(WebCore::AXRemoteFrame::generateRemoteToken const):
(WebCore::AXRemoteFrame::initializePlatformElementWithRemoteToken):
(WebCore::Accessibility::newAccessibilityRemoteToken):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityElements]):
(-[WebAccessibilityObjectWrapper accessibilityElementAtIndex:]):
(-[WebAccessibilityObjectWrapper _accessibilityRelativeFrame]):
(-[WebAccessibilityObjectWrapper accessibilityFrame]):
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/ios/AccessibilityIOS.h: Removed.
* Source/WebKit/Platform/ios/AccessibilityIOS.mm: Removed.
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _accessibilityRegisterUIProcessTokens]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::bindRemoteAccessibilityFrames):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::bindRemoteAccessibilityFrames):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::accessibilityRemoteTokenData const):
(WebKit::WebPage::updateRemotePageAccessibilityOffset):
(WebKit::WebPage::registerRemoteFrameAccessibilityTokens):
(WebKit::WebPage::createMockAccessibilityElement):
(WebKit::WebPage::accessibilityRemoteFrameOffset):
(WebKit::WebPage::accessibilityRemoteObject):

Canonical link: <a href="https://commits.webkit.org/288090@main">https://commits.webkit.org/288090@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/798f66fa1c30b163c404abfaa869bae8c158e7bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86273 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32725 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83835 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1288 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9075 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63763 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21487 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/935 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74378 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44049 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/835 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28552 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31178 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72219 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29154 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87712 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72100 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9154 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71329 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17794 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15415 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14334 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8920 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14452 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8761 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12284 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10569 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -171,6 +175,37 @@
+    m_processIdentifier = processIdentifier;
+
+    RetainPtr nsToken = toNSData(token);
+    NSDictionary *tokenDictionary = nsToken ? unarchivedTokenForData(nsToken).get() : nullptr;
```
Should this be `: nil` rather than `: nullptr` since this is assigning an `NSDictionary *`?

hoffmanjoshua replied:
Good point, I can change that!

twilco replied:
Noticed this one is marked resolved but it's still `nullptr` instead of `nil`, not sure if you got around to fixing this one yet though

hoffmanjoshua replied:
Somehow didn't staged + commit this one, sorry! Will be fixed in latest push

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -171,6 +175,37 @@
+    NSString *uuid = [tokenDictionary objectForKey:@"ax-uuid"];
+    AXRemoteElement* remoteElement = [[AXRemoteElement alloc] initWithUUID:uuid andRemotePid:processIdentifier andContextId:0];
+    remoteElement.onClientSide = YES;
+    remoteElement.accessibilityContainer = parentObject() ?  parentObject()->wrapper() : nil;
```
Should this be `parentObjectUnignored()`? That's what we use in `-[WebAccessibilityObjectWrapperIOS accessibilityContainer]`. But maybe this context is special?

Seems like there's an extra space after the `?`

hoffmanjoshua replied:
I used `parentObject()` here in case the scrollview/iframe was ignored, but that doesn't seem possible anymore. I will update this to use `parentObjectUnignored`.

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -171,6 +175,37 @@
+
+    m_remoteFramePlatformElement = adoptNS(remoteElement);
+
+    if (auto* cache = axObjectCache())
```
Probably want `CheckedPtr` rather than a raw pointer here?

fleizach commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -171,6 +176,37 @@
+        return;
+
+    NSString *uuid = [tokenDictionary objectForKey:@"ax-uuid"];
+    AXRemoteElement* remoteElement = [[AXRemoteElement alloc] initWithUUID:uuid andRemotePid:processIdentifier andContextId:0];
```
You're not using the soft linking here, which is trying to pull in the symbol

hoffmanjoshua replied:
That was my oversight, fixed in the latest commit!

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -171,6 +176,37 @@
+        return;
+
+    NSString *uuid = [tokenDictionary objectForKey:@"ax-uuid"];
+    AXRemoteElement* remoteElement = [allocAXRemoteElementInstance() initWithUUID:uuid andRemotePid:processIdentifier andContextId:0];
```
`AXRemoteElement` is an objective-C object, so should the `*` be by the `r` instead?

`AXRemoteElement *remoteElement`

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -171,6 +176,37 @@
+    NSString *uuid = [tokenDictionary objectForKey:@"ax-uuid"];
+    AXRemoteElement* remoteElement = [allocAXRemoteElementInstance() initWithUUID:uuid andRemotePid:processIdentifier andContextId:0];
+    remoteElement.onClientSide = YES;
+    remoteElement.accessibilityContainer = parentObjectUnignored() ?  parentObjectUnignored()->wrapper() : nil;
```
It would be a tiny bit more efficient to only get `parentObjectUnignored` only once, i.e.:

```cpp
RefPtr parent = parentObjectUnignored();
remoteElement.accessibilityContainer = parent ? parent->wrapper() : nil;
```

hoffmanjoshua committed:
AX: [Site Isolation] iOS remote element bring-up
https://bugs.webkit.org/show_bug.cgi?id=284847
rdar://problem/141640442

Reviewed by Tyler Wilcock.

This patch brings-up site isolation support for iOS accessibility. This involves
setting up AXRemoteFrames properly, using existing stub methods/IPC messages
that are used on macOS. After this, WebKit can create a remote element child
that binds the two processes with a matching token.

WebKit::newAccessibilityRemoteToken was moved to WebCore, so that it may be
used in AccessibilityObjectIOS.

Note: this patch does not address hit testing. That will be resolved in a follow
up.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/ios/AXRuntimeSPI.h: Added.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::addRemoteFrameChild):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::unarchivedTokenForData):
(WebCore::AXRemoteFrame::generateRemoteToken const):
(WebCore::AXRemoteFrame::initializePlatformElementWithRemoteToken):
(WebCore::Accessibility::newAccessibilityRemoteToken):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityElements]):
(-[WebAccessibilityObjectWrapper accessibilityElementAtIndex:]):
(-[WebAccessibilityObjectWrapper _accessibilityRelativeFrame]):
(-[WebAccessibilityObjectWrapper accessibilityFrame]):
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/ios/AccessibilityIOS.h: Removed.
* Source/WebKit/Platform/ios/AccessibilityIOS.mm: Removed.
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _accessibilityRegisterUIProcessTokens]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::bindRemoteAccessibilityFrames):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::bindRemoteAccessibilityFrames):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::accessibilityRemoteTokenData const):
(WebKit::WebPage::updateRemotePageAccessibilityOffset):
(WebKit::WebPage::registerRemoteFrameAccessibilityTokens):
(WebKit::WebPage::createMockAccessibilityElement):
(WebKit::WebPage::accessibilityRemoteFrameOffset):
(WebKit::WebPage::accessibilityRemoteObject):

Canonical link: https://commits.webkit.org/288090@main
